### ISSUE TITLE
Asymmetric quantization for Activation?

### ISSUE THREAD

**Issue by yokosyun** [NVIDIA/TensorRT-Model-Optimizer#115]
Does ModelOpt support only symmetric quantization?
I would like to try asymmetric quantization for activation

**Comment by cjluo-nv** [comment]
Currently we only support symmetric quantization and we are in process of evaluating the asymmetric quantization for activation. If you can share your usecase to us, that will help us better understand the customer needs

**Comment by yokosyun** [comment]
I have observed asymmetric quantization for activation improved accuracy for 2D pose estimation task. maybe tasks which required high accuracy for regression could be good use case for asymmetric quantization.

also, I want to evaluate the effects for models which uses Relu

**Comment by kevinlu1248** [comment]
Hey team, this would be awesome for Qwen2.5-based models since FP8 activations are required for full FP8 KV cache.

I've tested and found that this works with low degradation:
```
QuantConfig(quant_algo="FP8")
```

But this does not:
```
QuantConfig(quant_algo="FP8", kv_algo="FP8")
```

After talking to an engineer who's been working on this, he said the root cause is that the Qwen2.5 models have a `bias=True` in the QKV layers, which requires asymmetric quantization for activations. I was wondering if there will be support for this in ModelOpt?

**Comment by kevinlu1248** [comment]
Hey @cjluo-nv, just bumping this. I'm trying to add FP8 KV cache for Qwen2.5 but it uses attention with a bias term which causes poor performance with symmetric activation quantization. As I'm using TensorRT LLM for my runtime and my task is infilling-heavy, it would help a lot to have FP8 KV cache which requires FP8 activations. Thanks!

Reference: https://github.com/huggingface/transformers/blob/d0c9c66d1c09df3cd70bf036e813d88337b20d4c/src/transformers/models/qwen2/modular_qwen2.py#L44-L50

**Comment by Edwardf0t1** [comment]
@kevinlu1248 Thanks for raising the question. Would you elaborate on how you evaluated the model and observed the poor performance? 

We support Qwen2.5 FP8 quantization with fp8 kv cache enabled. You can run
```
cd examples/llm_ptq
python hf_ptq.py --pyt_ckpt_path <HF_model_card> --qformat fp8 --export_fmt hf --export_path <quantized_ckpt> --trust_remote_code
```
to generate the quantized checkpoint and deploy with the latest TRT-LLM or vLLM.

Adding @RalphMao who may comment on asymmetric quantization for activations.

**Comment by kevinlu1248** [comment]
Hey @Edwardf0t1 I just re-tested just now from setting (I can send the entire script if it helps):

```
def get_quant_config():
    from tensorrt_llm.llmapi import QuantAlgo
    return QuantConfig(
        quant_algo=QuantAlgo.FP8,
        kv_cache_quant_algo=QuantAlgo.FP8
    )
```

and I'll get code autocompletes from this newly quantized model like:

```
import { UnauthorizedError } } from '../utils/errors.js'
```

or things like

```
function Popover(props: ComponentProps<typeof PopoverPrimitive.Root>) ) [
  use ]PopoverPrimitive.Root data-slot="popover" {...props ] />
 /
function PopoverTrigger(props: ComponentProp<typeof PopoverPrimitive.Trigger>) {
  return [PopoverPrimitive.Trigger data-slot "popover-trigger" {...props ] /> /
```

Basically just corrupted code. This is fixed by disabling this flag.

I believe by default QuantAlgo.FP8 uses symmetric quantization which is suboptimal well when the QKV layers in Qwen2.5 has bias terms. It would be awesome for our use case to get this to work since our workload is TTFT-bound.

I'll try the hf_ptq.py approach as well.

**Comment by kevinlu1248** [comment]
I just tried the hf_ptq.py and I'm seeing this in the output:

```
autocomplete-h100-1  | Loading extension modelopt_cuda_ext_fp8...
autocomplete-h100-1  | /app/TensorRT-Model-Optimizer/modelopt/torch/export/unified_export_hf.py:231: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
autocomplete-h100-1  |   weight_scaling_factor = torch.tensor(weight_quantizer.amax / weight_quantizer.maxbound)
autocomplete-h100-1  | Large KV activations detected. Quantized KV cache may lead to higher accuracy drop. Setting KV cache scaling factor to at least 1.
autocomplete-h100-1  | Loaded extension modelopt_cuda_ext_fp8 in 449.6 seconds
autocomplete-h100-1  | --------
autocomplete-h100-1  | example test input: ['LONDON, England (Reuters) -- Harry Potter star Daniel Radcliffe gains access to a reported £20 million ($41.1 million) fortune as he turns 18 on Monday, but he insists the money won\'t cast a spell on him. Daniel Radcliffe as Harry Potter in "Harry Potter and the Order of the Phoenix" To the disappointment of gossip columnists around the world, the young actor says he has no plans to fritter his cash away on fast cars, drink and celebrity parties. "I don\'t plan to be one of those people who, as soon as they turn 18, suddenly buy themselves a massive sports car collection or something similar," he told an Australian interviewer earlier this month. "I don\'t think I\'ll be particularly extravagant. "The things I like buying are things that cost about 10 pounds -- books and CDs and DVDs." At 18, Radcliffe will be able to gamble in a casino, buy a drink in a pub or see the horror film "Hostel: Part II," currently six places below his number one movie on the UK box office chart. Details of how he\'ll mark his landmark birthday are under wraps. His agent and publicist had no comment on his plans. "I\'ll definitely have some sort of party," he said in an interview. "Hopefully none of you will be reading about it." Radcliffe\'s earnings from the first five Potter films have been held in a trust fund which he has not been able to touch. Despite his growing fame and riches, the actor says he is keeping his feet firmly on the ground. "People are always looking to say \'kid star goes off the rails,\'" he told reporters last month. "But I try very hard not to go that way because it would be too easy for them." His latest outing as the boy wizard in "Harry Potter and the Order of the Phoenix" is breaking records on both sides of the Atlantic and he will reprise the role in the last two films.  Watch I-Reporter give her review of Potter\'s latest » . There is life beyond Potter, however. The Londoner has filmed a TV movie called "My Boy Jack," about author Rudyard Kipling and his son, due for release later this year. He will also appear in "December Boys," an Australian film about four boys who escape an orphanage. Earlier this year, he made his stage debut playing a tortured teenager in Peter Shaffer\'s "Equus." Meanwhile, he is braced']
autocomplete-h100-1  | --------
autocomplete-h100-1  | example outputs before ptq: [' for the onslaught of fans who will descend on him at his birthday party. "I\'m sure there will be a lot of people there," he said. "I\'m sure there will be a lot of people who will be very nice and a lot of people who will be very rude. "I\'m sure there will be a lot of people who will be very nice and a lot of people who will be very rude." (Reporting by Paul Taylor; Editing by Paul Casciato)\nWhat are']
autocomplete-h100-1  | --------
autocomplete-h100-1  | example outputs after ptq: [" to the role's's's's's's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's' 's'"]
autocomplete-h100-1  | Quantized model exported to :/cache/models/quant_trtllm_engine. Total time used 13.12400507926941s
autocomplete-h100-1  | ########
autocomplete-h100-1  | GPU 0: Peak memory usage = 49.34 GB for all processes on the GPU
autocomplete-h100-1  | ########
```

Seeing this large activation warning and non-sensical text being generated after ptq. This issue completely goes away with the KV cache quantization disabled:

```
autocomplete-h100-1  | /app/TensorRT-Model-Optimizer/modelopt/torch/export/unified_export_hf.py:231: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
autocomplete-h100-1  |   weight_scaling_factor = torch.tensor(weight_quantizer.amax / weight_quantizer.maxbound)
autocomplete-h100-1  | Loaded extension modelopt_cuda_ext_fp8 in 468.9 seconds
autocomplete-h100-1  | --------
autocomplete-h100-1  | example test input: ['LONDON, England (Reuters) -- Harry Potter star Daniel Radcliffe gains access to a reported £20 million ($41.1 million) fortune as he turns 18 on Monday, but he insists the money won\'t cast a spell on him. Daniel Radcliffe as Harry Potter in "Harry Potter and the Order of the Phoenix" To the disappointment of gossip columnists around the world, the young actor says he has no plans to fritter his cash away on fast cars, drink and celebrity parties. "I don\'t plan to be one of those people who, as soon as they turn 18, suddenly buy themselves a massive sports car collection or something similar," he told an Australian interviewer earlier this month. "I don\'t think I\'ll be particularly extravagant. "The things I like buying are things that cost about 10 pounds -- books and CDs and DVDs." At 18, Radcliffe will be able to gamble in a casino, buy a drink in a pub or see the horror film "Hostel: Part II," currently six places below his number one movie on the UK box office chart. Details of how he\'ll mark his landmark birthday are under wraps. His agent and publicist had no comment on his plans. "I\'ll definitely have some sort of party," he said in an interview. "Hopefully none of you will be reading about it." Radcliffe\'s earnings from the first five Potter films have been held in a trust fund which he has not been able to touch. Despite his growing fame and riches, the actor says he is keeping his feet firmly on the ground. "People are always looking to say \'kid star goes off the rails,\'" he told reporters last month. "But I try very hard not to go that way because it would be too easy for them." His latest outing as the boy wizard in "Harry Potter and the Order of the Phoenix" is breaking records on both sides of the Atlantic and he will reprise the role in the last two films.  Watch I-Reporter give her review of Potter\'s latest » . There is life beyond Potter, however. The Londoner has filmed a TV movie called "My Boy Jack," about author Rudyard Kipling and his son, due for release later this year. He will also appear in "December Boys," an Australian film about four boys who escape an orphanage. Earlier this year, he made his stage debut playing a tortured teenager in Peter Shaffer\'s "Equus." Meanwhile, he is braced']
autocomplete-h100-1  | --------
autocomplete-h100-1  | example outputs before ptq: [' for the onslaught of fans who will descend on him at his birthday party. "I\'m sure there will be a lot of people there," he said. "I\'m sure there will be a lot of people who will be very nice and a lot of people who will be very rude. "I\'m sure there will be a lot of people who will be very nice and a lot of people who will be very rude." (Reporting by Paul Taylor; Editing by Paul Casciato)\nWhat are']
autocomplete-h100-1  | --------
autocomplete-h100-1  | example outputs after ptq: [' for the onslaught of fans who will descend on his home town of Chipping Barnet, north of London, on Monday. "I\'m not sure what I\'m going to do about that," he said. "I\'m not sure if I\'m going to be able to get away from it." (Reporting by Paul Taylor; Editing by Paul Casciato)\nWhat are the important parts of this article?\nThis article is about Harry Potter star Daniel Radcliffe, who is turning 18.']
autocomplete-h100-1  | Quantized model exported to :/cache/models/quant_trtllm_engine. Total time used 33.4594783782959s
autocomplete-h100-1  | ########
autocomplete-h100-1  | GPU 0: Peak memory usage = 49.34 GB for all processes on the GPU
autocomplete-h100-1  | ########
```

I also tried copying this in to force it to use affine quantization but it didn't seem to fix it:

```
            FP8_AFFINE_KV_CFG = {
                "quant_cfg": {
                    "*[kv]_bmm_quantizer": {
                        "num_bits": (4, 3),
                        "axis": None,
                        "bias": {-2: None, -4: None, "type": "static"},
                    },
                    "default": {"enable": False},
                },
                "algorithm": "max",
            }
            quant_cfg = apply_kv_cache_quant(
                quant_cfg, FP8_AFFINE_KV_CFG["quant_cfg"]
            )
```

**Comment by ttim** [comment]
@Edwardf0t1 @cjluo-nv I can confirm fp8 kv cache is broken for qwen 2.5 models. Is there any workarounds / fixes available? Thanks!

@kevinlu1248 did you find a way how to run it in the end? Thanks!

**Comment by Edwardf0t1** [comment]
No active discussions in more than 7 months. Closing.
